handlers/providers/azure.go

```diff
@@ -0,0 +1,54 @@
+package providers
+
+import (
+	b64 "encoding/base64"
+	"os"
+	"path/filepath"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancherio/go-rancher/client"
+)
+
+func init() {
+	if err := RegisterProvider("azure", azureHandler); err != nil {
+		log.Fatal("could not register azure provider")
+	}
+}
+
+func azureHandler(machine *client.Machine, machineDir string) error {
+	var data *string
+	var filename string
+	if machine.AzureConfig.SubscriptionCert != "" {
+		data = &machine.AzureConfig.SubscriptionCert
+		filename = "subscription-cert.pem"
+	} else {
+		data = &machine.AzureConfig.PublishSettingsFile
+		filename = "publish-settings.xml"
+	}
+	path, err := saveDataToFile(filename, *data, machineDir)
+	if err != nil {
+		return err
+	}
+	*data = path
```

Comment:
@wlan0 I dont think we can reassign the value of these fields (SubscriptionCert, PublishSettingsFile) like this. I think the original value should remain intact and that the file name should be set as a new data field. Note that the filename field doesn't need to be exposed to the user, it could just be an exposed property on the data field.

Reasons:
I realize that gms probably has some other idempotency problems, but doing this would cause more idempotency problems. Also, from an API perspective, it is not a good idea to change the value of a user-supplied field behind the scenes like this. Additionally, one of the things we'll need to do in the near future is make gms truly stateless so that it can be killed on host A and brought up on host B successfully.